tycho-core/src/main/java/org/eclipse/tycho/p2resolver/TargetDefinitionResolverService.java

```diff
@@ -117,6 +118,11 @@ private void debugCacheMiss(ResolutionArguments arguments) {
         }
     }
 
+    public URI resolveRepositoryLocation(String location) {
```

Comment:
Can we probably move the code for resolving variables from `TargetDefinitionResolver` into `TargetDefinitionResolverService` then? Also I think there should be exactly one method to replace variables, the parsing into an URL should happen at the relevant place, or we can have helper methods that call the generic one like resolveAsURI, resolveAsFile resolveAs...

Reply:
> Can we probably move the code for resolving variables from TargetDefinitionResolver into TargetDefinitionResolverService then?

Well yes, but then we have to pass `TargetDefinitionResolverService` into `TargetDefinitionResolver` as some sort of callback for resolving variables because the `resolve*` methods are needed deep in the code of `TargetDefinitionResolver`.
Maybe you had something like this in mind?
``` java
interface TargetDefinitionVariableResolver {
  String resolveVariable(String raw);
}

class TargetDefinitionResolverService implements TargetDefinitionVariableResolver{
  String resolveVariable(String raw){ 
    // move code for env_var, systep_prop and project_loc here
  }
  
  CompletableFuture<TargetDefinitionContent> resolveFromArguments(ResolutionArguments arguments) {
    ...
    var resolver = new TargetDefinitionResolver(this, arguments.environments, ...);
    resolver.resolveContent(...)
}

class TargetDefinitionResolver {
  final TargetDefinitionVariableResolver varResolver;
  TargetDefinitionResolver(TargetDefinitionVariableResolver varResolver, ...){}

  TargetDefinitionContent resolveContentWithExceptions(TargetDefinition definition, ...){
    ...
    } else if (locationDefinition instanceof PathLocation pathLocation) {
        String resolvePath = varResolver.resolvePath(pathLocation.getPath(), definition);
    ...
  }
}
```

personally I prefer having the resolution in `TargetDefinitionResolver` and delegating to it from `TargetDefinitionResolverService` where necessary but it is your call

> Also I think there should be exactly one method to replace variables,

ok I can do that, I did it as a separate method because I didn't want repository location to support `project_loc` as that makes no sense to me but I guess there is no harm in having that option available and just never user rather than creating special value for it

Reply:
Having a `TargetDefinitionVariableResolver` makes sense, maybe even as a separate component (it seems the mojos only reference `TargetDefinitionResolverService` because the actually want to resolve variables).

>  I did it as a separate method because I didn't want repository location to support project_loc as that makes no sense to me but I guess there is no harm in having that option available and just never user rather than creating special value for it

Yes we should just support "variables" the user has to decide what is the best one to use...

Reply:
What is the plexus way of doing `@PostConstruct`?

I created the `TargetDefinitionVariableResolver` but now I need to get the logger from `mavenContext` after injecting it
``` java
@Component(role = TargetDefinitionVariableResolver.class)
public class TargetDefinitionVariableResolver {

    @Requirement
    private MavenContext mavenContext;
    @Requirement // this does not work
    private MavenLogger logger;
    
    @PostConstruct // this does not work either
    private void postConstruct(){
      // I expect the DI framework to call this method after it injected all members
      this.logger = mavenContext.getLogger();
    }
```
of course I could get the logger at the start of the `resolve` method (so it sets the logger the first time it is called and does effectively noop on each subsequent call) but that seems ugly

Reply:
well both `org.eclipse.tycho.core.shared.MavenContext` and `org.eclipse.tycho.core.shared.MavenLogger` were used in the original `TargetDefinitionResolver`, I just moved the code from it into `TargetDefinitionVariableResolver` without doing any further changes.

If you wish I can replace the `logger` there by injected `org.codehaus.plexus.logging.Logger`.
As for the `mavenContext` it was used for
`mavenContext.getSessionProperties()` and `mavenContext.getProjects()` - what should I replace these by to avoid the `MavenContext`?

thanks for the post construct hint